test/integration_test_wrapper.py

```diff
@@ -0,0 +1,274 @@
+        run(['systemd-dissect', '--mount', image, td], check=True)
+        try:
+            yield td
+        finally:
+            log.debug(f"Unmounting {image}")
+            run(['systemd-dissect', '--umount', td], check=True)
+
+
+def main():
+    logging_basic_config(level=LOG_DEBUG)
+    args = parser.parse_args()
+
+    log.debug(f"test name: {args.test_name}\n"
+              f"image: {args.mkosi_image_name}\n"
+              f"mkosi output path: {args.mkosi_output_path}\n"
+              f"test unit name: {args.test_unit_name}\n"
+              f"mkosi args: {args.mkosi_args}\n"
+              f"hook module: {args.hook_module}")
+
+    per_test_output_dir = args.mkosi_output_path.parent / args.test_name
+    per_test_output_dir.parent.mkdir(parents=True, exist_ok=True)
+    log.debug(f"Copying {args.mkosi_output_path.absolute()} output dir "
+              f"to per-test path {per_test_output_dir.absolute()}")
+
+    hook = None
+    if args.hook_module is not None:
+        spec = spec_from_file_location('hook', args.hook_module)
+        hook = module_from_spec(spec)
+        sys_modules['hook'] = hook
+        spec.loader.exec_module(hook)
+
+    with ExitStack() as stack:
+        stack.enter_context(lock_file(per_test_output_dir))
+
+        with TemporaryDirectory(dir=per_test_output_dir.parent.absolute(),
+                                prefix=f".{per_test_output_dir.name}",
+                                suffix=".tmp") as td:
+            log.debug(f"Copying to temporary path {td}")
+            if geteuid() == 0:
+                cp_args = [
+                    'cp',
+                    '--archive',
+                    '--reflink=auto',
+                    '--no-target-directory',
+                    args.mkosi_output_path,
+                    td,
+                ]
+                if log.isEnabledFor(LOG_DEBUG):
+                    log.debug(
+                        "Using rootly copy of output directory with command "
+                        + ' '.join(quote(str(a)) for a in cp_args))
+                run(cp_args, check=True)
+            else:
+                log.debug("Using namespace aware copy of build directory")
+                for e in listdir(args.mkosi_output_path):
+                    src = args.mkosi_output_path / e
+                    dst = Path(td) / e
+                    cp_args = []
+
+                    if src.is_dir():
+                        cp_args += ['unshare', '--map-auto', '--map-root-user']
+
+                    args += ['cp', '--archive', '--reflink=auto', src, dst]
+                    run(cp_args, check=True)
+
+            log.debug(f"Cleaning old dir {per_test_output_dir.absolute()}")
+            with suppress(FileNotFoundError):
+                rmtree(per_test_output_dir)
+
+            log.debug(f"Renaming tempdir {td} onto "
+                      f"{per_test_output_dir.absolute()}")
+            rename(td, per_test_output_dir)
```

Comment:
All this can be replaced with mkosi's `--ephemeral` option which will do the copy for you. This also means we don't need a per test output directory anymore.